### PR TITLE
TINKERPOP-1063 TinkerGraph Performance

### PR DESCRIPTION
**Pull request by spmallette**
https://issues.apache.org/jira/browse/TINKERPOP-1063

Got rid of `Stream` usage in TinkerGraph vertex/edge iteration.  Tests all pass with `mvn clean install`.

VOTE +1

### CODE REVIEW

**Comment by okram** [comment]
Do we know that its faster?

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerGraph.java [hunk 1]
```diff
@@ -358,6 +350,14 @@ public Features features() {
...
     }
 
+    private void validateHomogenousIds(final List<Object> ids) {
+        final Class firstClass = ids.get(0).getClass();
+        for (Object id : ids) {
+            if (!id.getClass().equals(firstClass))
+                throw Graph.Exceptions.idArgsMustBeEitherIdOrElement();
+        }
+    }
+
```
Comment (dkuppitz) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerGraph.java]: This opens the door for NPE's.

```
gremlin> graph.vertices("a", null, 1)
java.lang.NullPointerException
Display stack trace? [yN] y
java.lang.NullPointerException
    at org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerGraph.validateHomogenousIds(TinkerGraph.java:356)
    at org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerGraph.createElementIterator(TinkerGraph.java:329)
    at org.apache.tinkerpop.gremlin.tinkergraph.structure.TinkerGraph.vertices(TinkerGraph.java:269)
```

Suggestion with a few tweaks:

```
private void validateHomogenousIds(final List<Object> ids) {
    final Iterator<Object> iterator = ids.iterator();
    Object id = iterator.next();
    if (id == null)
        throw Graph.Exceptions.idArgsMustBeEitherIdOrElement();
    final Class firstClass = id.getClass();
    while (iterator.hasNext()) {
        id = iterator.next();
        if (id == null || !id.getClass().equals(firstClass))
            throw Graph.Exceptions.idArgsMustBeEitherIdOrElement();
    }
}
```

**Comment by Jotschi** [comment]
@okram  ~18% to 16% faster without suggested tweaks (I used the Ferma Benchmark and compared merged PR with tp31 branch code)

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerGraph.java (hunk 1 above)
Comment (spmallette) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerGraph.java]: cool - i will integrate - nice find. might also add a test for null

**Comment by Jotschi** [comment]
Are TraversalStrategy optional? Or is it really needed to apply them? I just noticed that removing the TinkerGraphStepStrategy has a huge impact on the Ferma Benchmark results.

**Comment by spmallette** [comment]
If we drop the `TinkerGraphStepStrategy` then you lose index lookups and everything is a full scan.

**Comment by spmallette** [comment]
@dkuppitz I'd forgotten to implement that change you suggested on this one - all done now.

**Comment by dkuppitz** [comment]
VOTE: +1

**Comment by okram** [comment]
Cool stuff. `Stream` is evil. It lurks throughout the codebase... Its funny, we were all gun ho for Java8, but we barely use it :( .... VOTE +1.
